### PR TITLE
[SEQNG-1073] Display N&S offsets

### PR DESCRIPTION
**Pull request by rpiaggio**
image

### CODE REVIEW

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala [hunk 1]
```diff
@@ -21,32 +22,62 @@ object Formatting {
...
-    // Calculate the widest offset step
-    def sequenceOffsetWidths: (Double, Double) =
-      steps
-        .map(s =>
-          (tableTextWidth(offsetAngle(s.offsetP.toAngle)), tableTextWidth(offsetAngle(s.offsetQ.toAngle))))
-        .foldLeft((0.0, 0.0)) {
-          case ((p1, q1), (p2, q2)) => (p1.max(p2), q1.max(q2))
-        }
+    private def maxWidth(angles: List[Angle]): Double =
+      angles.map(angle => tableTextWidth(offsetAngle(angle))).max
```
[the remark this answers is not in the document]
Comment (swalker2m) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala]: There's `maximumOption` for `Foldable` instances.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/OffsetDisplayCell.scala [hunk 2]
```diff
@@ -38,49 +38,115 @@ object OffsetsDisplayCell {
...
+    )
+  }
+
+  private def nodAndShuffleOffsetsRender(step: NodAndShuffleStep, width: Double, axisLabelWidth: Double, nsNodLabelWidth: Double): VdomElement = {
+    val offsetBP = step.offset[OffsetType.NSNodB, Axis.P]
+    val offsetBQ = step.offset[OffsetType.NSNodB, Axis.Q]
+    val offsetAP = step.offset[OffsetType.NSNodA, Axis.P]
+    val offsetAQ = step.offset[OffsetType.NSNodA, Axis.Q]
+
+    <.table(
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/OffsetDisplayCell.scala]: I found it simpler to reason about, especially when I was playing with different possible ways of displaying all the offsets and labels.

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/OffsetDisplayCell.scala [hunk 3]
```diff
@@ -38,49 +38,115 @@ object OffsetsDisplayCell {
   private val guidingIcon = IconCrosshairs.copyIcon(color = "green".some, size = Size.Large)
   private val noGuidingIcon = IconBan.copyIcon(size = Size.Large)
 
+  private def standardOffsetsRender(step: StandardStep, offsetWidth: Double, axisLabelWidth: Double): VdomElement = {
+    val offsetP = step.offset[OffsetType.Telescope, Axis.P]
+    val offsetQ = step.offset[OffsetType.Telescope, Axis.Q]
+
+    <.table(
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/OffsetDisplayCell.scala]: Same as above.

File: modules/seqexec/model/src/main/scala/seqexec/model/Offset.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  implicit object OffsetAxisPFormat extends OffsetFormat[Axis.P] {
+    override val format = "p"
+  }
+
+  implicit object OffsetAxisQFormat extends OffsetFormat[Axis.Q] {
+    override val format = "q"
+  }
+}
+
+sealed trait OffsetConfigResolver[T, A] {
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply modules/seqexec/model/src/main/scala/seqexec/model/Offset.scala]: :)

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpiaggio) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala]: :+1:

File: modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala [hunk 5]
```diff
@@ -21,32 +22,62 @@ object Formatting {
...
+
+    // Calculate the widest offset step, widest axis label and widest NS nod label
+    def sequenceOffsetMaxWidth: (Double, Double, Double) =
+      steps.collect {
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply modules/seqexec/web/client/src/main/scala/seqexec/web/client/model/Formatting.scala]: I see what you mean. I'm not sure about the actual impact either.

We could try it out with a huge sequence.

**Comment by rpiaggio** [comment]
> This is looking very good, Perhaps @andrewwstephens wants to comment on the visuals
> I wonder about the use of `table` though. It seems to me these are not really tables and we could do the same using `div` and flexbox.

I'll attempt the `div`/`flex` approach.

**Pull request merged**